### DISCUSSION TITLE
How many threads does GO use?

### DISCUSSION THREAD

**Discussion by Penrhos1920** [GrandOrgue/grandorgue#1530] [Developers]
I have a multi core windows 10 machine with 32gb ram.  Yesterday I took a partial load of Buckeburg and changed it to load all samples at 16bit.  About 45 minutes later it was ready to play (actually ready for me to assign keyboards and stops etc).  During this time I had Task Manager open and could see that the cpu was only operating at 5% and none of the hard drives were doing much.  Memory varied from nearly 90% to 70%.

5% cpu is equivalent to about one thread on my 10 core Xeon.  Is GO only compiled for single thread?

PS. Nothing else was running except win10 background stuff.

**Comment by kerkovits** [comment]
GO is multi-threaded. The bottleneck is usually disk I/O while loading the samples for the first time.

**Reply by Penrhos1920** [reply]
The machine is an ex server so disk i/o shouldn't be a problem.  How do I investigate that?  Thanks

**Reply by kerkovits** [reply]
Do you have cache compression (File->Settings) enabled? That should be disabled on fast disks (i.e. SSD), if you have enough space. You may consider disable automatic caching at all. As far as I know, cache compression is implemented single-threaded.

**Comment by davidgritter** [comment]
The number of cores used for loading and for running can be separately defined in settings, see the grandorgue help file. I have found that even with SSDs using a SATA interface, the disk latency is the normal limiting factor, and use of a cache speeds things up.  However with an NVMe SSD, loading is much faster and a cache is a disadvantage (I set four cores for loading and eight for running.

**Reply by Penrhos1920** [reply]
Thanks.  I have all cores available for everything (after all the PC does nothing else!)

I've found loading a cached set the ssd only operates at 87%, the hdds 0, and cpu at 8%, I assume the limit is the motherboard ram access speed.

But when loading a set for the first time or reloading after changing an organ setting there is a long time when none of the drives are accessed and the cpu is only at 5-8% load.  The time varies between sets and Buckeburg is the longest by far at maybe 10 minute or so.
